src/Listeners/Plugin.php

```diff
@@ -44,7 +56,212 @@ public function deactivated( $plugin, $network_wide ) {
+		if ( 'plugin' !== $options['type'] ) {
+			return;
+		}
+		// set event type
+		if ( 'update' === $options['action'] ) {
+			$event_key = 'plugin_updated';
+
+			$plugins = [];
+			if ( isset( $options['plugins'] ) && is_array( $options['plugins'] ) ) {
+				foreach ( $options['plugins'] as $index => $pluginslug ) {
+					$plugin = $this->collect_plugin( 
+						$pluginslug, 
+						is_plugin_active( $pluginslug ),
+						false
+					);
+					array_push( $plugins, $plugin );
+				}
+			}
+			$data = array(
+				'plugins' => $plugins,
+			);
+		
+		} elseif ('install' === $options['action'] ) {
+			$event_key = 'plugin_installed';
+			// get all plugins - slug not returned for install actions
+			$plugins = $this->collect_plugins();
+			$data = array(
+				'plugins' => $plugins,
+			);
+
+		} else {
+			return;
+		}
+		
+		$this->push( $event_key, $data );
+	}
+
+	/**
+	 * Grab relevant data from plugin data - and only what we want
+	 * @param Array $data The plugin meta data from the header
+	 * @return Array Hiive relevant plugin details
+	 */
+	public static function glean_plugin_data( $data ){
+		$plugin = [];
+		$plugin['version'] = $data['Version'];
+		$plugin['title'] = $data['Name'] ? $data['Name'] : '';
+		$plugin['url'] = $data['PluginURI'] ? $data['PluginURI'] : '';
+		return $plugin;
+	}
+
+	/**
+	 * Get plugin dir name or main file if no dir. 
+	 * @param string  $slug Relative path to plugin file
+	 * @return string slug for plugin
+	 */
+	public static function glean_plugin_slugname( $slug ){
+		$newslug = '';
+		// if has `/` split on `/` and get second to last index.
+		if ( str_contains( $slug, '/' ) ) {
+			$parts = explode( '/', $slug );
+			end( $parts );
+			$newslug = prev( $parts );
+		} else {
+			// if not, split on . and get second to last index.
+			$parts = explode( '.', $slug );
+			end( $parts );
+			$newslug = prev( $parts );
+		}
+		return $newslug;
+	}
+
+	/**
+	 * Does this plugin autoupdate?
+	 * @param string  $slug plugin identifier
+	 * @return boolean Whether plugin is configured to auto-update
+	 */
+	public static function does_it_autoupdate( $slug ){
+		// check bluehost plugin setting for auto updates on all plugins
+		$bh_auto_updates = get_site_option( 'auto_update_plugin', 'true' );
+		if ( 'true' === $bh_auto_updates ) {
+			return true;
+		}
+
+		// check core setting for auto updates on this plugin
+		$wp_auto_updates = (array) get_site_option( 'auto_update_plugins', array() );
+		if ( in_array( $slug, $wp_auto_updates, true ) ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Prepare plugin data for single plugin 
+	 * For when single plugin action occurs
+	 * 
+	 * @param string  $slug Name of the plugin
+	 * @param boolean $active Whether the plugin is active
+	 * @param boolean $wrapper Whether to include a plugins wrapper array
+	 * @return Array of data for plugin 
+	 */
+	public static function collect_plugin( $slug, $active, $wrapper=true ){
```

Comment:
Why do we have the optional `$wrapper` paramater? It doesn't appear to be used anywhere and seems like it would make the return unreliable because the structure of where the plugin data is located would be different. 

I'm assuming we need the `$active` parameter because during the activation request `is_plugin_active()` won't return `true`? If not, then it seems like it would be better to just use `is_plugin_active()` inside this function and reduce the required parameters to just the slug.

Reply:
This is here because sometimes the context is different. I wanted to use one `collect_plugin` function. Sometimes it operates on individual items of an array vs a whole array of items and to use consistent data structure it had to wrap a single value into an array. This is due to the fact that WP deals differently when we're in an `upgrader_process_complete` handler . I should write up the distinctions better though because looking back at it now is confusing to me even. You can see the wrapper param set to false at https://github.com/bluehost/endurance-wp-module-data/pull/18/files/f9359a7ddf3f2cd6d3b986e25721471b4c7ec09e#diff-a916b4dbaac927a40a69935250a4d762f5926a5ba8fafe60d94e6dcb5fa9266cR106

And yea, the active param is because at the point we're hooking in it's not reliable what that will be so I'm explicitly setting it. IIRC the `plugin_deactivated` hook was returning a false positive there. I am using `is_plugin_active` in the upgrader handler though and is reliable there.

Reply:
Ok, how about instead of having this parameter, just have `collect_plugins()` call `collect_plugin()`? 

That way, the latter is what processes and gets the data about each plugin, but the former is what gets the whole group of them? 

Then when you need to supply an individual plugin in an array for the `activated` and `deactivated` hooks, you can use `array( $this->collect_plugin( $slug ) )`.

That will make the return types consistent and I thin will be easier to understand and troubleshoot. Also removes some duplicated code inside the two functions.

Reply:
Yea. Makes sense that a `collect_plugins` call would return an array and `collect_plugin` would return a single, and to use the second to provide data to the first. I'll work on that today.

Reply:
Though the use case for these is slightly different. The plural loads all plugins via `get_plugins`, and the singular version receives a slug. So not exactly apples to apples.

Reply:
Maybe `collect_installed_plugins()` then?

Reply:
Good function name. I've got it updated.